core/src/main/java/com/linecorp/armeria/client/Endpoint.java

```diff
@@ -418,21 +419,43 @@ public String authority() {
 
         if (isGroup()) {
             authority = "group:" + groupName;
-        } else if (port != 0) {
-            if (hostType == HostType.IPv6_ONLY) {
-                authority = '[' + host() + "]:" + port;
-            } else {
-                authority = host() + ':' + port;
-            }
-        } else if (hostType == HostType.IPv6_ONLY) {
-            authority = '[' + host() + ']';
         } else {
-            authority = host();
+            authority = buildAuthority(host());
         }
 
         return this.authority = authority;
     }
 
+    /**
+     * Converts this endpoint into a URI.
+     *
+     * @return the URI
+     */
+    public URI toURI(String scheme) {
+        requireNonNull(scheme, "scheme");
+
+        final String authority;
+        if (isGroup()) {
+            authority = "group:" + groupName;
+        } else {
+            authority = buildAuthority(hasIpAddr() ? ipAddr() : host());
+        }
+
+        return URI.create(scheme + "://" + authority);
+    }
+
+    private String buildAuthority(String host) {
+        if (NetUtil.isValidIpV6Address(host)) {
+            host = '[' + host + ']';
+        }
+
+        if (port != 0) {
+            host = host + ':' + port;
```

Comment:
Could you optimize this a little bit so that concatenation is not done in two-fold? e.g.

```java
if (port != 0) {
    if (isValidIpV6Address(host)) {
        return '[' + host + "]:" + port;
    } else {
        return host + ':' + port
    }
else {
    ...
}
```